### PR TITLE
[dv,cip] Mirror FATAL_ALERT_CAUSE-style registers in scoreboard

### PR DESCRIPTION
**Pull request by rswarbrick**
When running the TL error sequence run by `run_tl_intg_err_vseq_sub`, we
inject an integrity error into one or more TL accesses, which would
confuse the scoreboard for some blocks. To avoid chaos, we disable the
scoreboard when running this sequence with a runtime plusarg.

Some blocks' environments do prediction of hardware-updated CSRs, and
this is done in the scoreboard. Unfortunately, disabling the
scoreboard means that this won't happen which, in turn, causes
`csr_utils::csr_rd_check` to spot an "error" when reading a
hardware-updated CSR that changes on an alert (an "ALERT_CAUSE"
register).

This patch allows a block to configure the (single-bit) field that
will be set when there is a TL error. Assuming there's such a field
configured, `cip_base_scoreboard.sv` notices any error response and
updates the mirrored value of that field accordingly.

This allows us to get rid of the special-case code at the bottom of
`cip_base_vseq__tl_errors.svh`: rather than avoiding looking at a
particular behaviour, we can just predict it properly.

### CODE REVIEW

File: hw/ip/otbn/dv/uvm/env/otbn_env_cfg.sv
Comment (sriyerg) [inline hw/ip/otbn/dv/uvm/env/otbn_env_cfg.sv:56]: The design team is discussing on how to standardize this, so that per IP setting should not be needed. Could we hold off until then?

File: hw/dv/sv/cip_lib/cip_base_scoreboard.sv
Comment (sriyerg) [inline hw/dv/sv/cip_lib/cip_base_scoreboard.sv:407]: Same comment as before - these name lookups are expensive- if the design team can standardize this, then this logic can be simplified even more.

Comment (weicaiyang) [reply hw/dv/sv/cip_lib/cip_base_scoreboard.sv:407]: this function may make this simplier.
https://github.com/lowRISC/opentitan/blob/0b53455fe1ee60782cd40d5ff0fb4337ec5721c3/hw/dv/sv/str_utils/str_utils_pkg.sv#L93

**Review by weicaiyang — COMMENTED** [review]
thanks for adding this.
We probably have a more clear picture how design will standardize the alert cause next week. Then we know how we can handle it in DV side.

File: hw/ip/otbn/dv/uvm/env/otbn_env_cfg.sv
Comment (rswarbrick) [reply hw/ip/otbn/dv/uvm/env/otbn_env_cfg.sv:56]: At the moment, the relevant tests are failing for some seeds. I'd prefer to get something that works now. We can always remove the hook if everything ends up with the same name.

File: hw/dv/sv/cip_lib/cip_base_scoreboard.sv
Comment (rswarbrick) [reply hw/dv/sv/cip_lib/cip_base_scoreboard.sv:407]: @sriyerg: The name lookup only happens when we see an error transaction, so I wouldn't expect it to happen very often. Note that most blocks' scoreboards already do something like this twice (once for A and once for D) for every register transaction. I'd be astonished if this was measurable.

@weicaiyang: Ahah! I was sure there was something like that somewhere! I'll switch the code to use it.

File: hw/dv/sv/cip_lib/cip_base_scoreboard.sv
Comment (rswarbrick) [reply hw/dv/sv/cip_lib/cip_base_scoreboard.sv:93]: Ah, good point! I've changed this to:
```systemverilog
          // If we got an error response caused by an integrity failure, update the mirrored value
          // for any bus integrity alert field (if there is one).
          if (item.d_error &&
              cfg.en_tl_intg_gen &&
              !item.is_a_chan_intg_ok(.throw_error(0))) begin
            update_tl_alert_field_prediction();
          end
```
which I think is now right?

**Comment by weicaiyang** [comment]
Looks like it's not easy for designers to standardize the alert cause/error code for all the IPs. And This PR can handle the alert cause of TL integrity error for DV. 
I feel we could leave the design as it. Or if it's not hard to align, we can do some small change to unify the CSR name of alert cause. All use the same name either alert_cause or err_code to avoid confusion (unless they have different meaning). For how to code the field (one bit per error or not), we can leave it to block design owner.

@tjaychen

File: hw/dv/sv/cip_lib/cip_base_scoreboard.sv
Comment (sriyerg) [reply hw/dv/sv/cip_lib/cip_base_scoreboard.sv:407]: Logically, `ral.foo.bar.predict()` is going to be much faster than lookup up the register `foo` by name, then looking up the field `bar` by name. It of course, depends on how frequently is this logic exercized. Our runtimes are quite small, so we don't really notice it. We ought to change our scoreboards to not do the name lookup. That is unfortunately a limitation in the generated RAL model that duplicates identical CSR classes rather than defining the class once and instantiating an array of them. I am sure it will result in some degree of speedup - there's only way to find out what that going to look like.

File: hw/ip/otbn/dv/uvm/env/otbn_env_cfg.sv
Comment (sriyerg) [reply hw/ip/otbn/dv/uvm/env/otbn_env_cfg.sv:56]: Allright.

**Review by sriyerg — APPROVED** [review]

**Comment by sriyerg** [comment]
LGTM, pending Weicai's comments being addressed.

File: hw/dv/sv/cip_lib/cip_base_scoreboard.sv
Comment (rswarbrick) [reply hw/dv/sv/cip_lib/cip_base_scoreboard.sv:97]: Good idea. I've moved it and also updated the comment above `predict_tl_err` to explain exactly what the function does. This *does* mean that we only update things if `cfg.en_scb_tl_err_chk` but let's worry about that if it ever turns out to matter.

File: hw/dv/sv/cip_lib/cip_base_scoreboard.sv
Comment (rswarbrick) [reply hw/dv/sv/cip_lib/cip_base_scoreboard.sv:489]: Ah, sorry, I wasn't clear. I agree with what you're saying. I meant that I'd be astonished if this extra lookup was measurable. Maybe we should brainstorm a bit about ways we could make this more efficient: I'll open an issue in a minute.

**Review by rasmus-madsen — COMMENTED** [review]
LGTM

**Pull request merged**
